cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java

```diff
@@ -805,9 +805,25 @@ boolean isOpen() {
         Misc.checkNotNull(revisions, "Input revisions");
         Misc.checkArgument(!revisions.isEmpty(), "revisions cannot be empty");
 
-        // TODO hoist down queue.submit to just surround the call to RevsDiffBatchCallable
         try {
-            return get(queue.submit(new RevsDiffCallable(revisions)));
+
+            ValueListMap<String, String> missingRevs = new ValueListMap<String, String>();
+
+            // Break down by docId first to avoid potential rev ID clashes between doc IDs
+            for (Map.Entry<String, List<String>> entry : revisions.entrySet()) {
+                String docId = entry.getKey();
+                List<String> revs = entry.getValue();
+                // Partition into batches to avoid exceeding placeholder limit
+                // The doc ID will use one placeholder, so use limit - 1 for the number of
+                // revs for the remaining placeholders.
+                List<List<String>> batches = CollectionUtils.partition(revs,
+                        SQLITE_QUERY_PLACEHOLDERS_LIMIT - 1);
+
+                for (List<String> revsBatch : batches) {
+                    missingRevs.addValuesToKey(docId, get(queue.submit(new RevsDiffBatchCallable(docId, revsBatch))));
```

Comment:
For case 1, I think we can still get in the situation where the replication will error if a local doc with a matching id and rev gets written locally. Running the split on the queue does reduce the window for this to happen but could still easily happen if network latency is high enough. 

For case 2, since we only perform local revs diff when pulling, I don't think it is a problem right now, but with purge, I think that would come down to the exact use case, off the top of my head I can't think of a reason to purge a doc locally while it's still on the server and being actively pulled down.

We could submit all the calls at the same time and it should reduce the window for an operation to be interleaved with the calls for revs diff, but the exact gains would depend on the implementation of the executor service. It will need some work on the queue to enable submitting batches of tasks to be executed, but I think that is only a small amount of work.